### ISSUE TITLE
`exec` source: performance regression since 0.20.x

### ISSUE THREAD

**Issue by hhromic** [vectordotdev/vector#12515]
### A note for the community

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Problem

While benchmarking some of our pipelines before upgrading to Vector `0.21.1`, we noticed that the `exec` source seemed to have got a performance regression in `0.20.1` and `0.21.1` compared to `0.19.3`. :sad-noises:

It was suggested on Discord by the Vector team that the problem seems to be that `exec` does not currently support batching.

See the benchmarking details down below.

The following benchmarks measured the `exec` source while sending data to an `http` sink and a dummy HTTP server receiving and discarding the data from Vector. The source code for the dummy HTTP server is below:
```python
import asyncio
from aiohttp import web
import uvloop

async def handle(request):
    num_bytes = 0
    async for line in request.content:
        num_bytes += len(line)
    return web.Response(text=f"{num_bytes}\n")

app = web.Application()
app.add_routes([web.post('/', handle)])

if __name__ == '__main__':
    asyncio.set_event_loop_policy(uvloop.EventLoopPolicy())
    web.run_app(app)
```

The data used in the benchmarks is a file with `20000` real-world logs in plain-text. Some stats about this data:
```
average message size : 537.16 characters
minimum message size : 218 characters
maximum message size : 1277 characters
```

The following pipelines were used for benchmarking:

```yaml
# bl-exec.yaml - baseline with no transforms in the topology
sources:
  exec:
    type: exec
    command: ["cat", "data.txt"]
    mode: streaming
    streaming:
      respawn_on_exit: false
sinks:
  http:
    type: http
    inputs: ["exec"]
    uri: http://localhost:8080
    encoding:
      codec: ndjson
```
```yaml
# exec.yaml - a simple transform in the topology
sources:
  exec:
    type: exec
    command: ["cat", "data.txt"]
    mode: streaming
    streaming:
      respawn_on_exit: false
transforms:
  remap:
    type: remap
    inputs: ["exec"]
    drop_on_error: true
    source: |
      .parts = split!(.message, " ", limit: 2)
sinks:
  http:
    type: http
    inputs: ["remap"]
    uri: http://localhost:8080
    encoding:
      codec: ndjson
```

### Comparing `exec` baseline across Vector versions

```
$ hyperfine -w 3 -m 10 --export-markdown report.md \
  'vector-0.19.3/bin/vector -c bl-exec.yaml' \
  'vector-0.20.1/bin/vector -c bl-exec.yaml' \
  'vector-0.21.1/bin/vector -c bl-exec.yaml'
```
| Command | Mean [s] | Min [s] | Max [s] | Relative |
|:---|---:|---:|---:|---:|
| `vector-0.19.3/bin/vector -c bl-exec.yaml` | 1.212 ± 0.021 | 1.191 | 1.264 | 1.01 ± 0.02 |
| `vector-0.20.1/bin/vector -c bl-exec.yaml` | 1.237 ± 0.007 | 1.225 | 1.250 | 1.03 ± 0.01 |
| `vector-0.21.1/bin/vector -c bl-exec.yaml` | 1.205 ± 0.004 | 1.200 | 1.212 | 1.00 |

Not much difference in this experiment, `0.21.1` seems to perform slightly better than the rest and more stable.

### Comparing `exec` with transforms across Vector versions

```
$ hyperfine -w 3 -m 10 --export-markdown report.md \
  'vector-0.19.3/bin/vector -c exec.yaml' \
  'vector-0.20.1/bin/vector -c exec.yaml' \
  'vector-0.21.1/bin/vector -c exec.yaml'
```
| Command | Mean [s] | Min [s] | Max [s] | Relative |
|:---|---:|---:|---:|---:|
| `vector-0.19.3/bin/vector -c exec.yaml` | 1.275 ± 0.016 | 1.255 | 1.307 | 1.00 |
| `vector-0.20.1/bin/vector -c exec.yaml` | 1.293 ± 0.014 | 1.273 | 1.324 | 1.01 ± 0.02 |
| `vector-0.21.1/bin/vector -c exec.yaml` | 1.419 ± 0.007 | 1.405 | 1.425 | 1.11 ± 0.02 |

This time, `0.21.1` performed worse. The more transforms are in the topology, the larger the difference compared to `0.19.3`. For example, the same data but using a topology with two more transforms (that I cannot share unfortunately but are just a bunch of `replace()` and `parse_key_value()` in `remap`):

| Command | Mean [s] | Min [s] | Max [s] | Relative |
|:---|---:|---:|---:|---:|
| `vector-0.19.3/bin/vector -c vector.yaml` | 1.754 ± 0.032 | 1.719 | 1.827 | 1.00 |
| `vector-0.20.1/bin/vector -c vector.yaml` | 1.992 ± 0.061 | 1.898 | 2.089 | 1.14 ± 0.04 |
| `vector-0.21.1/bin/vector -c vector.yaml` | 2.801 ± 0.456 | 2.133 | 3.104 | 1.60 ± 0.26 |

For comparison, I performed the same experiment above but using `stdin` instead of `exec`:

| Command | Mean [s] | Min [s] | Max [s] | Relative |
|:---|---:|---:|---:|---:|
| `vector-0.19.3/bin/vector -c vector.yaml < data.txt` | 1.715 ± 0.018 | 1.690 | 1.749 | 1.00 |
| `vector-0.20.1/bin/vector -c vector.yaml < data.txt` | 1.919 ± 0.076 | 1.803 | 2.064 | 1.12 ± 0.05 |
| `vector-0.21.1/bin/vector -c vector.yaml < data.txt` | 1.983 ± 0.055 | 1.923 | 2.081 | 1.16 ± 0.03 |

While `0.19.3` is still the fastest, the difference is much less with the same topology and data.

### Version

vector 0.20.1 (x86_64-unknown-linux-gnu 38b806e 2022-04-07)
vector 0.21.1 (x86_64-unknown-linux-gnu 18787c0 2022-04-22)

### References

Might be related to #12504

**Comment by hhromic** [comment]
Got curious and made some benchmark using the `vm` runtime for the topology with transforms benchmark.

Using `exec` source:
| Command | Mean [s] | Min [s] | Max [s] | Relative |
|:---|---:|---:|---:|---:|
| `vector-0.21.1/bin/vector -c vector-ast.yaml` | 2.777 ± 0.477 | 2.069 | 3.103 | 1.00 |
| `vector-0.21.1/bin/vector -c vector-vm.yaml` | 2.889 ± 0.407 | 2.117 | 3.096 | 1.04 ± 0.23 |

Using `stdin` source:
| Command | Mean [s] | Min [s] | Max [s] | Relative |
|:---|---:|---:|---:|---:|
| `vector-0.21.1/bin/vector -c vector-ast.yaml < data.txt` | 1.993 ± 0.051 | 1.939 | 2.088 | 1.00 |
| `vector-0.21.1/bin/vector -c vector-vm.yaml < data.txt` | 2.021 ± 0.062 | 1.943 | 2.134 | 1.01 ± 0.04 |

The difference is not big (and not favorable to the VM runtime), however once again with `stdin` the difference is narrower.

**Comment by hhromic** [comment]
I ran this benchmark again using newer versions of Vector: `0.22.3`, `0.23.3`, `0.24.2` and `0.25.1` (latest).

In general, it seems that Vector got back to consistent performance since `0.22.x`. Great job, Vector team! 🎉
It can be clearly seen that `exec` did regress in `0.21.x` but it is now solved since `0.22.x`.

Therefore, I'm closing this issue now. 💯

---

### Comparing `exec` baseline (no transforms) across Vector versions

| Command | Mean [s] | Min [s] | Max [s] | Relative |
|:---|---:|---:|---:|---:|
| `/opt/vector-0.19.3/bin/vector -c bl-exec.yaml` | 1.233 ± 0.022 | 1.206 | 1.280 | 1.01 ± 0.02 |
| `/opt/vector-0.20.1/bin/vector -c bl-exec.yaml` | 1.255 ± 0.032 | 1.227 | 1.340 | 1.02 ± 0.03 |
| `/opt/vector-0.21.2/bin/vector -c bl-exec.yaml` | 1.226 ± 0.017 | 1.200 | 1.262 | 1.00 |
| `/opt/vector-0.22.3/bin/vector -c bl-exec.yaml` | 1.264 ± 0.027 | 1.235 | 1.317 | 1.03 ± 0.03 |
| `/opt/vector-0.23.3/bin/vector -c bl-exec.yaml` | 1.246 ± 0.041 | 1.218 | 1.359 | 1.02 ± 0.04 |
| `/opt/vector-0.24.2/bin/vector -c bl-exec-new.yaml` | 1.264 ± 0.040 | 1.216 | 1.337 | 1.03 ± 0.04 |
| `/opt/vector-0.25.1/bin/vector -c bl-exec-new.yaml` | 1.235 ± 0.019 | 1.214 | 1.267 | 1.01 ± 0.02 |

### Comparing `exec` with transforms across Vector versions

| Command | Mean [s] | Min [s] | Max [s] | Relative |
|:---|---:|---:|---:|---:|
| `/opt/vector-0.19.3/bin/vector -c exec.yaml` | 1.302 ± 0.041 | 1.251 | 1.363 | 1.03 ± 0.03 |
| `/opt/vector-0.20.1/bin/vector -c exec.yaml` | 1.330 ± 0.054 | 1.289 | 1.458 | 1.06 ± 0.04 |
| `/opt/vector-0.21.2/bin/vector -c exec.yaml` | 1.520 ± 0.085 | 1.422 | 1.667 | 1.21 ± 0.07 |
| `/opt/vector-0.22.3/bin/vector -c exec.yaml` | 1.338 ± 0.046 | 1.286 | 1.427 | 1.06 ± 0.04 |
| `/opt/vector-0.23.3/bin/vector -c exec.yaml` | 1.259 ± 0.014 | 1.234 | 1.283 | 1.00 |
| `/opt/vector-0.24.2/bin/vector -c exec-new.yaml` | 1.343 ± 0.065 | 1.297 | 1.516 | 1.07 ± 0.05 |
| `/opt/vector-0.25.1/bin/vector -c exec-new.yaml` | 1.374 ± 0.066 | 1.294 | 1.493 | 1.09 ± 0.05 |

### Comparing `stdin` with transforms across Vector versions

| Command | Mean [s] | Min [s] | Max [s] | Relative |
|:---|---:|---:|---:|---:|
| `/opt/vector-0.19.3/bin/vector -c stdin.yaml < data.txt` | 1.317 ± 0.044 | 1.263 | 1.407 | 1.08 ± 0.04 |
| `/opt/vector-0.20.1/bin/vector -c stdin.yaml < data.txt` | 1.324 ± 0.035 | 1.289 | 1.376 | 1.09 ± 0.03 |
| `/opt/vector-0.21.2/bin/vector -c stdin.yaml < data.txt` | 1.216 ± 0.019 | 1.196 | 1.250 | 1.00 |
| `/opt/vector-0.22.3/bin/vector -c stdin.yaml < data.txt` | 1.251 ± 0.025 | 1.226 | 1.311 | 1.03 ± 0.03 |
| `/opt/vector-0.23.3/bin/vector -c stdin.yaml < data.txt` | 1.228 ± 0.016 | 1.209 | 1.248 | 1.01 ± 0.02 |
| `/opt/vector-0.24.2/bin/vector -c stdin-new.yaml < data.txt` | 1.235 ± 0.017 | 1.205 | 1.256 | 1.01 ± 0.02 |
| `/opt/vector-0.25.1/bin/vector -c stdin-new.yaml < data.txt` | 1.264 ± 0.028 | 1.219 | 1.300 | 1.04 ± 0.03 |

### Comparing `exec` and `stdin` with transforms across Vector versions

| Command | Mean [s] | Min [s] | Max [s] | Relative |
|:---|---:|---:|---:|---:|
| `/opt/vector-0.19.3/bin/vector -c exec.yaml` | 1.360 ± 0.071 | 1.293 | 1.520 | 1.03 ± 0.06 |
| `/opt/vector-0.19.3/bin/vector -c stdin.yaml < data.txt` | 1.323 ± 0.027 | 1.283 | 1.379 | 1.00 |

| Command | Mean [s] | Min [s] | Max [s] | Relative |
|:---|---:|---:|---:|---:|
| `/opt/vector-0.20.1/bin/vector -c exec.yaml` | 1.370 ± 0.055 | 1.305 | 1.451 | 1.03 ± 0.05 |
| `/opt/vector-0.20.1/bin/vector -c stdin.yaml < data.txt` | 1.329 ± 0.033 | 1.281 | 1.376 | 1.00 |

| Command | Mean [s] | Min [s] | Max [s] | Relative |
|:---|---:|---:|---:|---:|
| `/opt/vector-0.21.2/bin/vector -c exec.yaml` | 1.524 ± 0.048 | 1.463 | 1.613 | 1.23 ± 0.05 |
| `/opt/vector-0.21.2/bin/vector -c stdin.yaml < data.txt` | 1.236 ± 0.037 | 1.198 | 1.298 | 1.00 |

| Command | Mean [s] | Min [s] | Max [s] | Relative |
|:---|---:|---:|---:|---:|
| `/opt/vector-0.22.3/bin/vector -c exec.yaml` | 1.329 ± 0.022 | 1.299 | 1.369 | 1.03 ± 0.04 |
| `/opt/vector-0.22.3/bin/vector -c stdin.yaml < data.txt` | 1.290 ± 0.047 | 1.242 | 1.383 | 1.00 |

| Command | Mean [s] | Min [s] | Max [s] | Relative |
|:---|---:|---:|---:|---:|
| `/opt/vector-0.23.3/bin/vector -c exec.yaml` | 1.281 ± 0.039 | 1.247 | 1.372 | 1.01 ± 0.05 |
| `/opt/vector-0.23.3/bin/vector -c stdin.yaml < data.txt` | 1.268 ± 0.055 | 1.220 | 1.368 | 1.00 |

| Command | Mean [s] | Min [s] | Max [s] | Relative |
|:---|---:|---:|---:|---:|
| `/opt/vector-0.24.2/bin/vector -c exec-new.yaml` | 1.330 ± 0.059 | 1.251 | 1.413 | 1.06 ± 0.05 |
| `/opt/vector-0.24.2/bin/vector -c stdin-new.yaml < data.txt` | 1.252 ± 0.020 | 1.227 | 1.291 | 1.00 |

| Command | Mean [s] | Min [s] | Max [s] | Relative |
|:---|---:|---:|---:|---:|
| `/opt/vector-0.25.1/bin/vector -c exec-new.yaml` | 1.297 ± 0.038 | 1.244 | 1.339 | 1.01 ± 0.04 |
| `/opt/vector-0.25.1/bin/vector -c stdin-new.yaml < data.txt` | 1.281 ± 0.033 | 1.243 | 1.354 | 1.00 |
